Added support for persistence with KV

larkin-nz opened the pull request:
Hey @hankjacobs,

I've added support for persistence via the API for the `KeyValueStore`.

When it is called with a `path` passed to the constructor it will create a file using the path specified.

Example usage:
```js
// Read and write from 'example.kvstore.db'
new KeyValueStore('example.kvstore.db')
```

I've added a test also so you can validate it all works correctly.

I'd also appreciate it if you could release it in a new release on NPM as soon as possible as I'd like to be using this functionality with Restt-CLI.

Keep up the great work :)

hankjacobs commented on lib/kv.js:
```diff
@@ -1,12 +1,35 @@
 
 class KeyValueStore {
-  constructor () {
+  constructor (name) {
```
Rather than `name`, let's call this `path` and have it be the absolute or relative path to the file. This will simplify the logic below. We will leave it to the caller to decide on the extension and where the file get's stored.

hankjacobs replied:
Additionally, if `path` has a length of zero, is undefined or null persistence should be considered disabled.

larkin-nz replied:
I've updated to use a path now

hankjacobs commented on lib/__tests__/kv.test.js:
```diff
@@ -79,4 +80,22 @@ describe('kv', () => {
     expect(() => { kv.delete('hello') }).toThrowError(new Error('HTTP DELETE request failed: 404 Not Found'))
   })
+
+  test('KeyValueStore persistence', async () => {
```
For async tests, [accept a callback function](https://github.com/dollarshaveclub/cloudworker/blob/29b36611fcb211dfadfecf995e62f78c40406f06/lib/__tests__/cloudworker-e2e.test.js#L39) as a parameter and [invoke it after the test assertion(s](https://github.com/dollarshaveclub/cloudworker/blob/29b36611fcb211dfadfecf995e62f78c40406f06/lib/__tests__/cloudworker-e2e.test.js#L55)

larkin-nz replied:
Done this also

hankjacobs commented on lib/__tests__/kv.test.js:
```diff
@@ -79,4 +80,22 @@ describe('kv', () => {
   })
+
+  test('KeyValueStore persistence', async () => {
+    // Create a persistent store
```
Could clean up most of these comments? I'd prefer comments be saved for code that is unclear / could be misunderstood. Variable assignments and code that makes sense by reading it shouldn't have comments.

hankjacobs commented on README.md:
```diff
@@ -35,13 +35,14 @@ cw.dispatch(req).then((res) => {
+  -p, --port <port>                   Port (default: 3000)
+  -d, --debug                         Debug
+  -s, --kv-set [variable.key=value]   Binds variable to a local implementation of Workers KV and sets key to value (default: [])
+  -f, --kv-file [variable=path]       Set the filepath for value peristence for the local implementation of Workers KV (defualt: [])
```
typo "defualt"

larkin-nz replied:
Corrected

larkin-nz commented:
Hey @hankjacobs - added features as requested.

I've added the flag `--kv-set` a replacement for `--set` with a deprecation warning logged if `--set` is used.

I've added the flag `--kv-file` to use as the file path for KV persistence.

hankjacobs commented on lib/utils.js:
```diff
@@ -15,13 +15,42 @@ module.exports.extractKVBindings = (setFlagValues) => {
+
+    if (setFlagPaths) {
+      // Find a variable path from the paths which matches the variable
+      const variablePath = setFlagPaths.find(value => {
```
Let's build an object of variables to paths outside of this for loop. We can do all the format checking there rather than in two places (here and below). Then all that's needed here is a lookup into that object.

larkin-nz replied:
Totally agree - was rushed earlier

hankjacobs commented on lib/utils.js:
```diff
@@ -15,13 +15,42 @@ module.exports.extractKVBindings = (setFlagValues) => {
   }
 
+  if (setFlagPaths.length > 0 && setFlagValues.length == 0) {
+    for (const flag of setFlagPaths) {
```
Use the object based on my comment above here instead of parsing the flag paths again.

larkin-nz replied:
Good idea - resolved now

hankjacobs commented:
Oh and don't forget to update with master

larkin-nz commented:
@hankjacobs 

All patched and optimised now :)

hankjacobs commented on lib/utils.js:
```diff
@@ -16,15 +17,37 @@ module.exports.extractKVBindings = (setFlagValues) => {
   }
 
+  for (const [variable, path] of Object.entries(filepaths)) {
+    bindings[variable] = new KeyValueStore(path)
```
I believe this should only be done if a binding for the variable doesn't exist yet. I could be wrong but as it's currently written, all variables that have key-values set and have a file path will be overwritten with an empty KeyValueStore.

larkin-nz replied:
Yes, good point - this was supposed to come before the other loop!

hankjacobs commented on lib/utils.js:
```diff
@@ -1,12 +1,13 @@
 const {KeyValueStore} = require('./kv')
 module.exports.read = f => fs.readFileSync(f).toString('utf-8')
-module.exports.extractKVBindings = (setFlagValues) => {
+module.exports.extractKVBindings = (kvSetFlags, kvFileFlags) => {
```
Sorry I missed this on the first CR but please add tests for the new argument and logic.

larkin-nz replied:
Added testing now

hankjacobs commented on lib/utils.js:
```diff
@@ -16,15 +17,37 @@ module.exports.extractKVBindings = (setFlagValues) => {
+const extractKVPaths = (kvFileFlags) => {
+  const paths = {}
+
+  if (kvFileFlags && kvFileFlags.length > 0) {
```
length check is redundant. Furthermore, I'm a fan of returning early if no work needs to be done. Rather than wrapping the for loop in a conditional, let's put something like the below before the for loop:
```
if (!kvFileFlags) { return paths }
```

larkin-nz replied:
Yes, good call, resolved!

larkin-nz commented:
@hankjacobs 

All resolved now - this should be it!

hankjacobs commented on lib/utils.js:
```diff
@@ -25,6 +32,26 @@ module.exports.extractKVBindings = (setFlagValues) => {
+
+  if (!kvFileFlags) return paths
+
+  if (kvFileFlags.length > 0) {
```
This is unnecessary.

larkin-nz replied:
Corrected

larkin-nz commented:
@hankjacobs 

Yes, corrected now!

hankjacobs commented:
Merged and beta pushed with it included. I'll play with it tomorrow and publish a new version if it looks good.

larkin-nz commented:
Excellent - thank you!

hankjacobs commented:
@larkin-nz v0.0.11 has been released

larkin-nz commented:
Excellent - thank you!
